### DISCUSSION TITLE
Can't fetch menuitems from Wordpress with GraphQL and Apolloclient

### DISCUSSION THREAD

**Discussion by diceboi** [vercel/next.js#45424] [Help] [answered]
Hi, I'm pretty new in webdev, just learning nextjs. I try to solve this issue from several days now but somehow its just didn't work for me.
I looked up multiple forums, tutorial videos, everything, but can't find what I did wrong.

I try to fetch my Primary Nav from Wordpress via GraphQL, and it looks like nothing has returned.

I created a Mainnav, and a Secondarynav components, and imported them into my Headernav component. In the page.js I also imported my Headernav just to quickly wrap up everything.

In my Mainnav where I try to fetch data from wordpress, that looks like this:

`
```
import { AiOutlineCreditCard } from 'react-icons/ai'
import { IoIosSearch, IoIosMenu } from 'react-icons/io'
import { gql } from "@apollo/client";
import { client } from "@/app/apollo-client";
import Link from 'next/link';

export default function MainNav({menus = []}) {
    return (
        <>
        <div className="flex flex-wrap justify-center px-4 w-full h-24 mx-auto">
            <div className='flex items-center gap-8 w-10/12 h-16'>
                <div id="logo" className="flex shrink-0 items-center w-40 h-14">
                    <a href="#">
                        <img src="https://teszt.arpusz.hu/wp-content/uploads/2023/01/ph-logo.svg" alt="logo" className="w-40" />
                    </a>
                </div>
                <div id="search" className="flex shrink items-center w-full h-14">
                    <div className='relative flex items-center justify-center w-full bg-neutral-100 h-10 rounded-full hover:bg-neutral-300'>
                        <input type="text" className="border-0 rounded-full w-full h-10 bg-neutral-100 px-4 hover:bg-neutral-300 focus:outline-none focus:bg-neutral-300 ease-in-out duration-300"/>
                        <IoIosSearch className='absolute right-3 h-6 w-auto'/>                        
                    </div>                   
                </div>
                <div id="proficlub" className="flex items-center h-14 self-end">
                    <button className="flex flex-nowrap items-center gap-3 bg-rose-600 hover:bg-rose-700 ease-in-out duration-300 h-10 px-2.5 rounded-full text-white"><AiOutlineCreditCard /> ProfiClub</button>
                </div>
            </div>
            <div className='flex items-center w-10/12 h-14 border-b border-slate-200'>
                <ul id="menu" className='flex items-center gap-8 text-base font-bold h-14'>
                    {
                        menus.map((menu) => {
                            return (
                                    <li>                       
                                        <Link href={(menu.uri)} key={menu.id} >{menu.label}</Link>
                                    </li>
                                    )
                        })                    
                    }                    
                </ul>
                <div id="userbuttons" className=''></div>
            </div>
        </div>
        </>
    )
  }


  export async function getStaticProps () {
    const GET_MENUS = gql`
    query PrimaryMenuItems {
    menuItems(where: {location: PRIMARY}) {
        nodes {
            label
            uri
            id
            }
        }
    }
    `;

    const response = await client.query({
        query: GET_MENUS,
    })

    const menus = response?.data?.menuItems?.nodes;

    return {
        props: {
            menus,
        }
    }

};
`
```

![image](https://user-images.githubusercontent.com/122260041/215630456-d2de763b-f956-4b07-a159-4cf3cf72387b.png)

The Nav should be here, but inside my <ul></ul> elements there is nothing shows up. Its completely empty in the DevTool.

Every little help would be appriciated! Thanks!

Public repo: https://github.com/diceboi/profihome-nextjs

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You've got a few concepts mixed up. Please read carefully.

As of Next 13+, there are two methods to build applications:

- `pages`
- `appDir` using a folder named `app`

While it is possible for a project to make use of both, it is not possible to interoperate between both. I mean that you cannot mix up methods that were used in `pages` into `app`.

Formally, a lot of the data fetching done in `pages` has changed in `app`. 

Since you are just starting and your snippet contains `getStaticProps`, I'd say that you best forget about `app` for this particular application. 

So you action point number 1 is to get rid of `app` in your https://github.com/diceboi/profihome-nextjs/tree/main/src root.

Action point number 2 would be to have a file `pages/_app.jsx`, and also create `pages/index.jsx` which would be the home page.

You also need to move your styling and perhaps merge the `Head` into `_app`'s rendering.

I think by now you'll be pretty overwhelmed, but we are nearly there.

One of the various reasons to create the `appDir` technology into Next.js was that `pages` is very restrictive when it comes to server data fetching. Functions such as `getServerSideProps` and `getStaticProps` can only exist at page level. That is `pages/index.jsx` or perhaps `pages/users/dashboard.jsx`. 

In https://github.com/diceboi/profihome-nextjs/blob/main/src/components/headernav/mainnav.jsx, you are using `getStaticProps`, and that's fundamentally wrong, and won't work, because this is a component.

> Depending on your level of expertise, you know there's a way to import these into a page and re-export them, but I don't want to overwhelm OP more than necessary.

That has to be placed, for example, at `pages/index.jsx`, and passed through which ever technique, down to MainNav. Render-props, context, redux, SWR, react-query, etc, there's various ways to do this.

This is what's wrong with your application. On top of that you threw in `app`, and imported that component into https://github.com/diceboi/profihome-nextjs/blob/main/src/app/page.js, which is already a non-starter, `getStaticProps` will never run under these conditions. 

I am not going to explore how to do this within `app` because this is already pretty long, and the current state of `appDir` requires some kind of experience building applications.

**Reply by diceboi** [reply]
Thank you for your answer! To be honest i didnt know that. The reason why I go for the app way is to be up to date with next. I will remove it now and start learning the routs I can import getStaticProps to components as you said. If I have any trouble I may write it here, if you are ok with it. Thank you again, I may never find that out by myself :)

**Reply by icyJoseph** [reply]
Go ahead! looking forward!
